File: sdk/AuthSys-SWIFT-Example/src/AuthSysClient.swift
import Foundation

class AuthSysClient {
    private let appSecret: String
    private let version: String
    private let apiURL: String
    private let session: URLSession

    var sessionToken = ""
    var lastError = ""
    var lastResponse = ""
    var initialized = false
    var username = ""
    var email = ""
    private var variables: [String: Any] = [:]

    init(appSecret: String, version: String,
         apiURL: String = "https://authsys-main-production.up.railway.app/api/v1") {
        self.appSecret = appSecret
        self.version = version
        var url = apiURL
        while url.hasSuffix("/") { url = String(url.dropLast()) }
        self.apiURL = url
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    private func getHWID() -> String {
        var raw = ""

        #if os(Windows)
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "wmic")
        task.arguments = ["csproduct", "get", "uuid"]
        #elseif os(macOS)
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/usr/sbin/ioreg")
        task.arguments = ["-rd1", "-c", "IOPlatformExpertDevice"]
        #else
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/bin/cat")
        task.arguments = ["/etc/machine-id"]
        #endif

        let pipe = Pipe()
        task.standardOutput = pipe
        do {
            try task.run()
            task.waitUntilExit()
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            if let output = String(data: data, encoding: .utf8) {
                raw = output.trimmingCharacters(in: .whitespacesAndNewlines)

                #if os(Windows)
                let lines = raw.components(separatedBy: "\n")
                if lines.count >= 2 { raw = lines[1].trimmingCharacters(in: .whitespacesAndNewlines) }
                #elseif os(macOS)
                for line in raw.components(separatedBy: "\n") {
                    if line.contains("IOPlatformUUID") {
                        let parts = line.components(separatedBy: "\"")
                        if parts.count >= 4 { raw = parts[3]; break }
                    }
                }
                #endif
            }
        } catch {}

        if raw.isEmpty { raw = Host.current().name ?? "unknown" }
        return raw
    }

    private func getJSON(_ key: String, from jsonString: String) -> String {
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let val = json[key] else { return "" }
        if let s = val as? String { return s }
        if let n = val as? NSNumber { return n.stringValue }
        if let d = try? JSONSerialization.data(withJSONObject: val) {
            return String(data: d, encoding: .utf8) ?? ""
        }
        return ""
    }

    private func post(endpoint: String, body: [String: Any], token: String = "") async -> String {
        guard let url = URL(string: "\(apiURL)/client/\(endpoint)") else {
            return "{\"success\":false,\"detail\":\"Invalid URL\"}"
        }
        var req = URLRequest(url: url, timeoutInterval: 30)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !token.isEmpty {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            req.setValue(getHWID(), forHTTPHeaderField: "X-HWID")
        }
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (data, _) = try await session.data(for: req)
            return String(data: data, encoding: .utf8) ?? ""
        } catch {
            return "{\"success\":false,\"detail\":\"\(error.localizedDescription)\"}"
        }
    }

    func initApp(appName: String) async {
        lastError = ""
        lastResponse = ""
        initialized = false

        let body: [String: Any] = [
            "app_secret": appSecret,
            "version": version,
            "app_name": appName,
            "hwid": getHWID()
        ]
        lastResponse = await post(endpoint: "init", body: body)

        let status = getJSON("status", from: lastResponse)
        if status == "success" || status == "update_available" {
            initialized = true
            let varsStr = getJSON("variables", from: lastResponse)
            if !varsStr.isEmpty, let data = varsStr.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                variables = json
            }
        } else {
            lastError = getJSON("detail", from: lastResponse)
            if lastError.isEmpty { lastError = "Init failed" }
        }
    }

    func login(username: String, password: String, sessionLength: Int = 86400) async {
        sessionToken = ""
        lastError = ""
        lastResponse = ""

        let body: [String: Any] = [
            "app_secret": appSecret,
            "username": username,
            "password": password,
            "hwid": getHWID(),
            "session_length": sessionLength
        ]
        lastResponse = await post(endpoint: "login", body: body)

        let detail = getJSON("detail", from: lastResponse)
        if !detail.isEmpty { lastError = detail; return }

        let success = getJSON("success", from: lastResponse)
        if success == "true" {
            sessionToken = getJSON("token", from: lastResponse)
            self.username = username
            email = getJSON("email", from: lastResponse)
        } else {
            lastError = "Login failed"
        }
    }

    func register(username: String, password: String, licenseKey: String, email: String = "") async {
        lastError = ""
        lastResponse = ""

        var body: [String: Any] = [
            "app_secret": appSecret,
            "username": username,
            "password": password,
            "license_key": licenseKey,
            "hwid": getHWID()
        ]
        if !email.isEmpty { body["email"] = email }
        lastResponse = await post(endpoint: "register", body: body)

        let detail = getJSON("detail", from: lastResponse)
        if !detail.isEmpty { lastError = detail; return }

        let success = getJSON("success", from: lastResponse)
        if success != "true" { lastError = "Registration failed" }
    }

    func licenseLogin(licenseKey: String, sessionLength: Int = 86400) async {
        sessionToken = ""
        lastError = ""
        lastResponse = ""

        let body: [String: Any] = [
            "app_secret": appSecret,
            "license_key": licenseKey,
            "hwid": getHWID(),
            "session_length": sessionLength
        ]
        lastResponse = await post(endpoint: "license-login", body: body)

        let detail = getJSON("detail", from: lastResponse)
        if !detail.isEmpty { lastError = detail; return }

        let success = getJSON("success", from: lastResponse)
        if success == "true" {
            sessionToken = getJSON("token", from: lastResponse)
            username = getJSON("username", from: lastResponse)
        } else {
            lastError = "License login failed"
        }
    }

    func licenseCheck(licenseKey: String) async {
        lastError = ""
        lastResponse = ""
        let body: [String: Any] = [
            "app_secret": appSecret,
            "license_key": licenseKey
        ]
        lastResponse = await post(endpoint: "license/check", body: body)
    }

    func verify() async {
        lastError = ""
        lastResponse = ""
        if sessionToken.isEmpty {
            lastError = "No active session"
            return
        }
        lastResponse = await post(endpoint: "verify", body: [:], token: sessionToken)
    }

    func chatSend(roomId: Int, message: String) async {
        lastError = ""
        lastResponse = ""
        if sessionToken.isEmpty {
            lastError = "No active session"
            return
        }
        let encMsg = message.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? message
        lastResponse = await post(endpoint: "chat/send?room_id=\(roomId)&message=\(encMsg)", body: [:], token: sessionToken)
    }

    func var(_ name: String) -> String {
        if let val = variables[name] {
            if let s = val as? String { return s }
            if let d = try? JSONSerialization.data(withJSONObject: val) {
                return String(data: d, encoding: .utf8) ?? ""
            }
        }
        return ""
    }

    func logout() {
        sessionToken = ""
        username = ""
        email = ""
        lastError = ""
        lastResponse = ""
    }
}

File: sdk/AuthSys-KOTLIN-Example/src/AuthSysClient.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class AuthSysClient(
    private val appSecret: String,
    private val version: String,
    private val apiURL: String = "https://authsys-main-production.up.railway.app/api/v1"
) {
    var sessionToken = ""
        private set
    var lastError = ""
        private set
    var lastResponse = ""
        private set
    var initialized = false
        private set
    var username = ""
        private set
    var email = ""
        private set
    private var variables: MutableMap<String, Any> = mutableMapOf()

    private val baseUrl: String = apiURL.trimEnd('/')

    private fun getHWID(): String {
        try {
            val os = System.getProperty("os.name").lowercase()
            val process = when {
                os.contains("win") -> Runtime.getRuntime().exec(
                    arrayOf("wmic", "csproduct", "get", "uuid"))
                os.contains("mac") -> Runtime.getRuntime().exec(
                    arrayOf("/usr/sbin/ioreg", "-rd1", "-c", "IOPlatformExpertDevice"))
                else -> Runtime.getRuntime().exec(
                    arrayOf("cat", "/etc/machine-id"))
            }
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val output = reader.readText().trim()
            reader.close()
            process.waitFor()

            return when {
                os.contains("win") -> {
                    val lines = output.split("\n")
                    if (lines.size >= 2) lines[1].trim() else output
                }
                os.contains("mac") -> {
                    for (line in output.split("\n")) {
                        if (line.contains("IOPlatformUUID")) {
                            val parts = line.split("\"")
                            if (parts.size >= 4) return parts[3]
                        }
                    }
                    output
                }
                else -> output
            }
        } catch (_: Exception) {}
        return java.net.InetAddress.getLocalHost().hostName
    }

    private fun getJSON(key: String, json: String): String {
        return try {
            val obj = org.json.JSONObject(json)
            if (obj.has(key)) obj.get(key).toString() else ""
        } catch (_: Exception) { "" }
    }

    private fun post(endpoint: String, body: Map<String, Any>, token: String = ""): String {
        return try {
            val url = URL("$baseUrl/client/$endpoint")
            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.doOutput = true
            conn.connectTimeout = 30000
            conn.readTimeout = 30000

            if (token.isNotEmpty()) {
                conn.setRequestProperty("Authorization", "Bearer $token")
                conn.setRequestProperty("X-HWID", getHWID())
            }

            val jsonBody = org.json.JSONObject(body).toString()
            OutputStreamWriter(conn.outputStream).use { it.write(jsonBody) }

            val status = conn.responseCode
            val reader = BufferedReader(InputStreamReader(
                if (status in 200..299) conn.inputStream else conn.errorStream
            ))
            val resp = reader.readText()
            reader.close()
            conn.disconnect()
            resp
        } catch (e: Exception) {
            """{"success":false,"detail":"${e.message}"}"""
        }
    }

    fun initApp(appName: String) {
        lastError = ""
        lastResponse = ""
        initialized = false

        val body = mapOf<String, Any>(
            "app_secret" to appSecret,
            "version" to version,
            "app_name" to appName,
            "hwid" to getHWID()
        )
        lastResponse = post("init", body)

        val status = getJSON("status", lastResponse)
        if (status == "success" || status == "update_available") {
            initialized = true
            val varsStr = getJSON("variables", lastResponse)
            if (varsStr.isNotEmpty()) {
                try {
                    val jsonObj = org.json.JSONObject(varsStr)
                    for (key in jsonObj.keySet()) {
                        variables[key] = jsonObj.get(key)
                    }
                } catch (_: Exception) {}
            }
        } else {
            lastError = getJSON("detail", lastResponse)
            if (lastError.isEmpty()) lastError = "Init failed"
        }
    }

    fun login(username: String, password: String, sessionLength: Int = 86400) {
        sessionToken = ""
        lastError = ""
        lastResponse = ""

        val body = mapOf<String, Any>(
            "app_secret" to appSecret,
            "username" to username,
            "password" to password,
            "hwid" to getHWID(),
            "session_length" to sessionLength
        )
        lastResponse = post("login", body)

        val detail = getJSON("detail", lastResponse)
        if (detail.isNotEmpty()) { lastError = detail; return }

        val success = getJSON("success", lastResponse)
        if (success == "true") {
            sessionToken = getJSON("token", lastResponse)
            this.username = username
            email = getJSON("email", lastResponse)
        } else {
            lastError = "Login failed"
        }
    }

    fun register(username: String, password: String, licenseKey: String, email: String = "") {
        lastError = ""
        lastResponse = ""

        val body = mutableMapOf<String, Any>(
            "app_secret" to appSecret,
            "username" to username,
            "password" to password,
            "license_key" to licenseKey,
            "hwid" to getHWID()
        )
        if (email.isNotEmpty()) body["email"] = email
        lastResponse = post("register", body)

        val detail = getJSON("detail", lastResponse)
        if (detail.isNotEmpty()) { lastError = detail; return }

        val success = getJSON("success", lastResponse)
        if (success != "true") lastError = "Registration failed"
    }

    fun licenseLogin(licenseKey: String, sessionLength: Int = 86400) {
        sessionToken = ""
        lastError = ""
        lastResponse = ""

        val body = mapOf<String, Any>(
            "app_secret" to appSecret,
            "license_key" to licenseKey,
            "hwid" to getHWID(),
            "session_length" to sessionLength
        )
        lastResponse = post("license-login", body)

        val detail = getJSON("detail", lastResponse)
        if (detail.isNotEmpty()) { lastError = detail; return }

        val success = getJSON("success", lastResponse)
        if (success == "true") {
            sessionToken = getJSON("token", lastResponse)
            username = getJSON("username", lastResponse)
        } else {
            lastError = "License login failed"
        }
    }

    fun licenseCheck(licenseKey: String) {
        lastError = ""
        lastResponse = ""
        val body = mapOf<String, Any>(
            "app_secret" to appSecret,
            "license_key" to licenseKey
        )
        lastResponse = post("license/check", body)
    }

    fun verify() {
        lastError = ""
        lastResponse = ""
        if (sessionToken.isEmpty()) {
            lastError = "No active session"
            return
        }
        lastResponse = post("verify", emptyMap(), sessionToken)
    }

    fun chatSend(roomId: Int, message: String) {
        lastError = ""
        lastResponse = ""
        if (sessionToken.isEmpty()) {
            lastError = "No active session"
            return
        }
        val encMsg = URLEncoder.encode(message, "UTF-8")
        lastResponse = post("chat/send?room_id=$roomId&message=$encMsg", emptyMap(), sessionToken)
    }

    fun `var`(name: String): String {
        return variables[name]?.toString() ?: ""
    }

    fun logout() {
        sessionToken = ""
        username = ""
        email = ""
        lastError = ""
        lastResponse = ""
    }
}
